File: src/registry/persist.ts
import { join } from "node:path";
import { MenvError } from "../core/errors.ts";
import { writeFileAtomic } from "../io/write.ts";
import type { Registry } from "./types.ts";
import { validateRegistry } from "./validate.ts";

export const REGISTRY_FILENAME = "menv.json";

export async function loadRegistry(root: string): Promise<Registry> {
  const file = Bun.file(join(root, REGISTRY_FILENAME));
  if (!(await file.exists())) {
    throw new MenvError("NOT_FOUND", `no ${REGISTRY_FILENAME} found in ${root} — run \`menv init\` first`);
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(await file.text());
  } catch (e) {
    throw new MenvError("PARSE", `${REGISTRY_FILENAME} is not valid JSON: ${(e as Error).message}`);
  }
  const { registry, issues } = validateRegistry(parsed);
  if (registry === null) {
    const summary = issues.map((i) => `${i.path}: ${i.message}`).join("; ");
    throw new MenvError("VALIDATION", `${REGISTRY_FILENAME} is invalid — ${summary}`, issues);
  }
  return registry;
}

// Canonical on-disk form: 2-space indent + trailing newline, so diffs stay
// minimal and generated edits are byte-stable.
export async function saveRegistry(root: string, registry: Registry): Promise<void> {
  await writeFileAtomic(root, REGISTRY_FILENAME, `${JSON.stringify(registry, null, 2)}\n`);
}

File: go/internal/registry/persist.go
package registry

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/nikrabaev/menv/internal/core"
	menvio "github.com/nikrabaev/menv/internal/io"
)

const RegistryFilename = "menv.json"

// LoadRegistry reads menv.json from root, parses, and validates it.
func LoadRegistry(root string) (Registry, error) {
	path := filepath.Join(root, RegistryFilename)
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return Registry{}, &core.MenvError{
				Code:    core.ErrNotFound,
				Message: fmt.Sprintf("no %s found in %s — run `menv init` first", RegistryFilename, root),
			}
		}
		return Registry{}, &core.MenvError{
			Code:    core.ErrVaultIO,
			Message: fmt.Sprintf("could not read %s: %v", RegistryFilename, err),
		}
	}

	var raw Registry
	if err := json.Unmarshal(data, &raw); err != nil {
		return Registry{}, &core.MenvError{
			Code:    core.ErrParse,
			Message: fmt.Sprintf("%s is not valid JSON: %v", RegistryFilename, err),
		}
	}

	issues := ValidateRegistry(raw)
	if len(issues) > 0 {
		parts := make([]string, len(issues))
		for i, iss := range issues {
			parts[i] = iss.String()
		}
		return Registry{}, &core.MenvError{
			Code:    core.ErrValidation,
			Message: fmt.Sprintf("%s is invalid — %s", RegistryFilename, strings.Join(parts, "; ")),
			Details: issues,
		}
	}

	return raw, nil
}

// SaveRegistry writes the registry to menv.json with canonical formatting:
// 2-space indent + trailing newline, so diffs stay minimal and byte-stable.
func SaveRegistry(root string, r Registry) error {
	data, err := json.MarshalIndent(r, "", "  ")
	if err != nil {
		return &core.MenvError{
			Code:    core.ErrVaultIO,
			Message: fmt.Sprintf("could not serialize registry: %v", err),
		}
	}
	content := append(data, '\n')
	return menvio.WriteFileAtomic(root, RegistryFilename, content)
}
